File: trie/trie.py
class Node:
    def __init__(self, flag=False):
        self.flag = flag
        self.children = [None] * 26

    def isEnd(self) -> bool:
        return self.flag

    def setEnd(self) -> None:
        self.flag = True

    def containsKey(self, ch: str):
        return self.children[ord(ch) - ord("a")] != None

    def putKey(self, ch: str, node):
        self.children[ord(ch) - ord("a")] = node

    def getKey(self, ch):
        return self.children[ord(ch) - ord("a")]


class Trie:
    def __init__(self):
        self.root = Node()

    def insert(self, word: str):
        temp = self.root
        for ch in word:
            if temp is not None:
                if temp.containsKey(ch) == False:
                    temp.putKey(ch, Node())
                temp = temp.getKey(ch)
        assert temp is not None
        temp.setEnd()

    def search(self, word: str):
        temp = self.root
        for ch in word:
            if temp is not None:
                if not temp.containsKey(ch):
                    return False
                temp = temp.getKey(ch)
        assert temp is not None
        return temp.isEnd()

    def startsWith(self, word: str):
        temp = self.root

        for ch in word:
            if temp is not None:
                if not temp.containsKey(ch):
                    return False
                temp = temp.getKey(ch)
        return True

File: trie/trie.cpp
#include <iostream>
#include <string>

class Node {
public:
  bool mFlag;
  Node *mChildren[26];
  Node() : mFlag{false} {}
  Node(bool flag) : mFlag{flag} {}

  bool containsKey(char ch) {
    int idx = int(ch) - (int)'a';
    return mChildren[idx] != nullptr;
  }

  void put(char ch, Node *node) {

    int idx = int(ch) - (int)'a';
    mChildren[idx] = node;
  }

  Node *get(char ch) {

    int idx = int(ch) - (int)'a';
    return mChildren[idx];
  }

  void setEnd() { mFlag = true; }

  bool isEnd() { return mFlag; }
};

class Trie {
  Node *root;

public:
  void insert(std::string word) {

    Node *node = root;
    for (char &ch : word) {
      if (!node->containsKey(ch)) {
        node->put(ch, new Node);
      }
      node = node->get(ch);
    }
    node->setEnd();
  }

  bool search(std::string word) {
    Node *node = root;
    for (char &ch : word) {
      if (node->containsKey(ch)) {
        node = node->get(ch);
      } else {
        return false;
      }
    }
    return node->isEnd();
  }

  bool startsWith(std::string word) {
    Node *node = root;
    for (char &ch : word) {
      if (node->containsKey(ch)) {
        node = node->get(ch);
      } else {
        return false;
      }
    }
    return true;
  }
};

int main(int argc, char *argv[]) {


  return 0;
}
